server/hwapi/endpoints/certification/logic.py

```diff
@@ -46,27 +51,203 @@ def is_certified(system_info: CertificationStatusRequest, db: Session):
     kernel = report.kernel
     bios = report.bios
 
-    if latest_certificate:
-        return True, CertifiedResponse(
-            status=CertificationStatus.CERTIFIED,
-            os=data_validators.OSValidator(
-                distributor="Canonical Ltd.",
-                version=latest_certificate.release.release,
-                codename=latest_certificate.release.codename,
-                kernel=data_validators.KernelPackageValidator(
-                    name=kernel.name, version=kernel.version, signature=kernel.signature
-                ),
-                loaded_modules=[],
+    return True, CertifiedResponse(
+        os=data_validators.OSValidator(
+            distributor="Canonical Ltd.",
+            version=latest_certificate.release.release,
+            codename=latest_certificate.release.codename,
+            kernel=data_validators.KernelPackageValidator(
+                name=kernel.name, version=kernel.version, signature=kernel.signature
             ),
-            bios=(
-                data_validators.BiosValidator(
-                    release_date=bios.release_date,
-                    revision=bios.revision,
-                    firmware_revision=bios.firmware_revision,
-                    vendor=bios.vendor.name,
-                    version=bios.version,
-                )
-                if bios
-                else None
-            ),
-        )
+            loaded_modules=[],
+        ),
+        bios=(
+            data_validators.BiosValidator(
+                release_date=bios.release_date,
+                revision=bios.revision,
+                firmware_revision=bios.firmware_revision,
+                vendor=bios.vendor.name,
+                version=bios.version,
+            )
+            if bios
+            else None
+        ),
+    )
+
+
+def is_partially_certified(
+    system_info: CertificationStatusRequest, db: Session
+) -> tuple[bool, RelatedCertifiedSystemExistsResponse | None]:
+    """
+    Check if some components of the system were seen on other certified machines.
+    """
+    release = (
+        repository.get_release_from_os(db, system_info.os) if system_info.os else None
+    )
+    board = repository.get_board_by_validator_data(db, system_info.board)
+    machines = repository.get_machines_with_same_hardware_params(
+        db, system_info.architecture, board, release
+    )
+    machine_ids = [machine.id for machine in machines]
```

Comment:
small point here, but unless we specifically need a list then making a generator would be more memory efficient

Reply:
this logic has been removed, mark as resolved